torch/utils/benchmark/utils/timer.py

```diff
@@ -21,6 +22,57 @@ def timer() -> float:
     timer = timeit.default_timer
 
 
+PYTHON = "Python"
+CPP = "C++"
+
+
+class CPPTimer:
+    def __init__(
+        self,
+        stmt: str,
+        setup: str,
+        timer: Callable[[], float],
+        globals: Dict[str, Any],
+    ) -> None:
+        if timer is not timeit.default_timer:
+            # FIXME: We should be able to automatically detect GPU operations.
+            raise ValueError(
+                "When timing C++, timer must be `timeit.default_timer`. If you "
+                "did not specify `timer=...`, this means that a timer with "
+                "CUDA syncronization was chosen, and you should pass "
+                "`timer=timeit.default_timer` to the Timer's constructor to "
+                "declare that CUDA syncronization is not needed."
+            )
+
+        if len(globals) > 1 or "torch" not in globals:
+            # Timer will add `torch` as a convenience.
+            raise ValueError("C++ timing does not support globals.")
```

Comment:
If I somehow explicitly pass an empty globals, won't this still raise a value error? That doesn't seem right.

Reply:
The whole concept of globals here doesn't make sense (the torch global doesn't make sense either, the corresponding C++ concept doesn't really have any relation to what you're doing here.) Seems better to just not have globals at all in this case.

Reply:
This was working around the fact that there was a convenience `globals.set_default("torch", torch)`. A cleaner solution is to just separate Python specific and C++ specific logic in `__init__` and simplify downstream checks, which I've done.